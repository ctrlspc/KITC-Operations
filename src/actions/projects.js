import database from '../firebase/firebase';

export const setProjects = (projects) => ({
  type: 'SET_PROJECTS',
  projects
});

export const startSetProjects = () => {
  return (dispatch) => {
    return database.ref(`projects`)
      .once('value')
      .then((snapshot) => {
        const projects = [];

        snapshot.forEach(childsnapshot => {
          projects.push({
            id:childsnapshot.key,
            ...childsnapshot.val()
          });
        });        
        dispatch(setProjects(projects));
      });
  };
};

export const addProject = (project) => ({
  type: 'ADD_PROJECT',
  project
});

export const startAddProject = ({title, description, projectManager, projectType}) => {
 return (dispatch) => {
    const project = {title, description, projectManager, projectType};

    return database.ref(`projects`)
      .push(project)
      .then((ref) => {
        dispatch(addProject({
          id:ref.key,
          ...project
        }));

        return Promise.resolve(ref.key);
      });
  };
};

export const updateProject = (id, updates) => ({
  type: 'UPDATE_PROJECT',
  id, 
  updates
});

export const startUpdateProject = (id, updates) => {
  return (dispatch) => {
    return database.ref(`projects/${id}`)
      .update(updates)
      .then((ref) => {
        dispatch(updateProject(id, updates));
      });
    };
};
