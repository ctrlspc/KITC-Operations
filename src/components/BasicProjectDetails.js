import React from 'react';

export default ({title, description, projectManager, projectType}) => (
  <div>
    <h3>{title}</h3>
    <p>{description}</p>
    <p><span>Project Manager:</span> {projectManager.name}</p>
  </div>
);