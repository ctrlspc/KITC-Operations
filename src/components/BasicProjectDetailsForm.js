import React from 'react';

export default class BasicProjectDetailsForm extends React.Component {
 
  constructor(props) {
    super(props);
    const project = this.props.project;
    this.state = {
      title: project ? project.title : '',
      description: project ? project.description : '',
      projectManager: project ? project.projectManager  : this.props.projectManagers[0] ,
      projectType: project ? project.projectType : 'ext',
      error:''
    };
  };

  onTitleChange = (e) => {
    const title = e.target.value;
    this.setState(() => ({title}));
  };

  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({description}));
  };

  onManagerChange = (e) => {
    const projectManager = this.props.projectManagers.find((projectManager) => {
      return projectManager.id === e.target.value
    });
    this.setState(() => ({projectManager}));
  };

  onTypeChange = (e) => {
    const projectType = e.target.value;
    this.setState(() => ({projectType}));
  };

  onSubmit = (e) => {
    e.preventDefault();

    if(!this.state.title || !this.state.description || !this.state.projectManager) {
      this.setState(() => ({error:'Please provide a title and project description'}));
    } else {
      this.setState(() => ({error:''}));
      this.props.onSubmit({
        title: this.state.title,
        description: this.state.description,
        projectManager: this.state.projectManager ,
        projectType: this.state.projectType
      });
    }
  };

  render () {
    return (
      <form className="form" onSubmit={this.onSubmit}>
        {this.state.error && <p className="form__error">{this.state.error}</p> }
        <input 
          autoFocus
          className="text-input"
          type="text" 
          placeholder="Project Name"
          value={this.state.title}
          onChange={this.onTitleChange}
        />
        <textarea 
          className="textarea"
          placeholder="Project Description"
          value={this.state.description}
          onChange={this.onDescriptionChange}
        />
        <div className="select-group">
          <div className="select-group__item">
            <h3 className="select-group__title">Project Manager</h3>
            <select 
              className="select-group__select"
              id="project-manager"
              onChange={this.onManagerChange}
              value={!!this.state.projectManager ? this.state.projectManager.id : ''}
            >
              {this.props.projectManagers && this.props.projectManagers.map((item) => {
                return <option key={item.id} value={item.id}>{item.name}</option> 
              })}
            </select>
          </div>
          <div className="select-group__item">
            <h3 className="select-group__title">Project Type</h3>
            <select 
              className="select-group__select"
              id="project-type" 
              onChange={this.onTypeChange}
              value={this.state.projectType}
            >
              <option value="int">Internal</option>
              <option value="ext">External</option>
            </select>
          </div>
        </div>
        <div>
          <button className="button">Save Project</button>
        </div>
        
      </form>
    );
  };
};
