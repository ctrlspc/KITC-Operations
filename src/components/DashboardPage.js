import React from 'react';
import ProjectList from './ProjectList';
import { Link } from 'react-router-dom';

const DashboardPage = () => (

  <div>
    <div className="page-header">
      <div className="content-container">
        <h1 className="page-header__title">
          Projects Dashboard
        </h1>
        <div className="page-header__actions">
          <Link className="button" to="/create">Create a Project</Link>
        </div>
      </div>
    </div>
    <div className="content-container">
      <ProjectList projects={[
        {
          id:'123',
          title:'A project',
          description:'a description',
          projectManager:{id:'1',name:'Jason Marshall'},
          projectType:'int'
        }
      ]} />
    </div>
  </div>
  
);

export default DashboardPage;