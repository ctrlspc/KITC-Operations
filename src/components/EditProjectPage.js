import React from 'react';
import { connect } from 'react-redux';
import ProjectForm from '../components/ProjectForm';
import { updateProject } from '../actions/projects';

export class EditProjectPage extends React.Component {

  onSubmit = (project) => {
    this.props.updateProject(this.props.project.id, project);
    this.props.history.push('/');
  }

  render () {
    return (
      <div className="content-container">
        <h1>Edit an existing Project</h1>
        <ProjectForm 
          onSubmit={this.onSubmit} 
          projectManagers={this.props.projectManagers}
          project={this.props.project}
        />
      </div>
    );
  };
};

export const mapStateToProps = (state, props) => ({
  project: state.projects.find((project) => project.id === props.match.params.id),
  projectManagers: state.projectManagers
})

export const mapDispatchToProps = (dispatch) => ({
  updateProject: (id, project) => dispatch(updateProject(id, project))
})


export default connect(mapStateToProps,mapDispatchToProps)(EditProjectPage);