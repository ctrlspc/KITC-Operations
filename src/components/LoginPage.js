import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from '../actions/iam';

export const LoginPage = ({ startLogin }) => (
  <div className="box-layout">
    <div className="box-layout__box">
      <h1 className="box-layout__title">KITC</h1>
      <p>Project Operations Tracker</p>
      <button className="button" onClick={startLogin}>Login with Google</button>
    </div>
    
  </div>
);

export const mapDispatchToProps = (dispatch) => ({
  startLogin: () => dispatch(startLogin())
});

export default connect(undefined, mapDispatchToProps)(LoginPage);