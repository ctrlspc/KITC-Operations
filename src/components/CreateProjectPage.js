import React from 'react';
import { connect } from 'react-redux';

import BasicProjectDetailsForm from './BasicProjectDetailsForm';
import { startAddProject } from '../actions/projects';
import { getActiveUsers } from '../reducers';
import { startSetUsers } from '../actions/users';

export class CreateProjectPage extends React.Component {

  constructor(props) {
    super(props);
    this.props.getUsers();
  }

  onSubmit = (project) => {
    this.props.addProject(project);
    this.props.history.push('/');
  };

  render () {
    return (
      <div className="content-container">
        <h1>Create a new Project</h1>
        <BasicProjectDetailsForm 
          onSubmit={this.onSubmit} 
          projectManagers={this.props.users}
        />
      </div>
    );
  };
};

export const mapDispatchToProps = (dispatch) => ({
  getUsers: () => dispatch(startSetUsers()),
  addProject: (project) => dispatch(startAddProject(project))
});

export const mapStateToProps = (state) => ({
  users: getActiveUsers(state)
});

export default connect(mapStateToProps, mapDispatchToProps)(CreateProjectPage);


