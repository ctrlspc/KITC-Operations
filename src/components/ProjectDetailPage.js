import React from 'react';
import { connect } from 'react-redux';
import BasicProjectDetails from './BasicProjectDetails';
import BasicProjectDetailsForm from './BasicProjectDetailsForm';
import { startUpdateProject } from '../actions/projects';
import { getActiveUsers } from '../reducers';
import { startSetUsers } from '../actions/users';

export class ProjectDetailPage extends React.Component {

  constructor(props) {
    super(props);
    this.props.getUsers();
    this.state = {
      editBasicProjectDetails: !!props.editBasicProjectDetails
    };
  }

  onEditBasicProjectDetails = () => {
    this.setState(() => ({editBasicProjectDetails: true}));
  };

  onFinishedEditingBasicProjectDetails = (updates) => {
    this.props.updateProject(this.props.project.id, updates).then(() => {
      this.setState(() => ({editBasicProjectDetails: false}));
    });
  }

  render (){
    return (
      <div className="content-container">
        <div className="box-section">
          <div className="box-section__header">
            <h1>Basic Project Details</h1>
            {!this.state.editBasicProjectDetails && 
              <button className="button" onClick={this.onEditBasicProjectDetails}>Edit</button>
            }
          </div>
          <div className="box-section__body">
            {
              this.state.editBasicProjectDetails ? 
              (
                <BasicProjectDetailsForm 
                  onSubmit={this.onFinishedEditingBasicProjectDetails} 
                  projectManagers={this.props.users}
                  project={this.props.project}
                />
              ) : (
                <BasicProjectDetails {...this.props.project} onEditClick={this.onEditBasicProjectDetails}/>
              )
            }
          </div>
        </div>
      </div>
    );
  };
};

export const mapStateToProps = (state, props) => ({
  project: state.projects.find((project) => project.id === props.match.params.id),
  users: getActiveUsers(state)
});

export const mapDispatchToProps = (dispatch) => ({
  getUsers: () => dispatch(startSetUsers()),
  updateProject: (id, project) => dispatch(startUpdateProject(id, project))
})

export default connect(mapStateToProps, mapDispatchToProps)(ProjectDetailPage);