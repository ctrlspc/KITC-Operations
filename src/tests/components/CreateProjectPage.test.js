import React from 'react';
import { shallow } from 'enzyme';
import { CreateProjectPage } from '../../components/CreateProjectPage';
import projects from '../fixtures/projects';

let addProject, history, wrapper;

beforeEach(() => {
  addProject = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(<CreateProjectPage addProject={addProject} history={history}/>);
});

test('should render the CreateProjectPage component correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should handle onSubmit', () => {
  wrapper.find('ProjectForm').prop('onSubmit')(projects[0]);
  expect(addProject).toHaveBeenLastCalledWith(projects[0]);
  expect(history.push).toHaveBeenLastCalledWith('/');
})