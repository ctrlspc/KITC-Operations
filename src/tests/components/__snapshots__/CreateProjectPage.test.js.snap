// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the CreateProjectPage component correctly 1`] = `
<div
  className="content-container"
>
  <h1>
    Create a new Project
  </h1>
  <BasicProjectDetailsForm
    onSubmit={[Function]}
  />
</div>
`;
